docs: entra id configuration using terraform (#57835)
gravitational/teleport: docs/pages/identity-governance/entra-id/terraform.mdx

---
title:  Configure the Entra ID Integration using Terraform
description: Describes how to set up the Teleport Entra ID integration using Terraform
sidebar_position: 3
labels:
 - how-to
 - identity-governance
 - terraform
---

> This guide shows end-to-end Entra ID integration configuration using Terraform and `tctl`.

Reviewer: I would add a sentence describing what the integration entails. What functionality does it add?

> This guide shows end-to-end Entra ID integration configuration using Terraform and `tctl`.

Author: Is that necessary? This guide is already under Entra ID section and our docs navigation is good enough for user to get more context on what piece of content they are reading both in-product navigated and search landing users. 
Instead of repeating such detail in every guide, its generally better to point them to the root doc and have a canonical reference. That helps both for readers and us maintainers.  

<img width="285" height="303" alt="Screenshot 2025-08-25 at 3 39 14 PM" src="https://github.com/user-attachments/assets/7d953fbd-b612-4a98-848a-c2d17653a5fe" />

Feel free to convince me otherwise :)

You will configure Entra ID side of configuration using Terraform and install the Entra ID 
plugin in Teleport using `tctl`.

See [getting started with Entra ID integration](getting-started.mdx) for a guided end-to-end Entra ID configuration. 

## Prerequisites

- Teleport Identity Governance enabled for your Teleport cluster.
- Your user must have privileged administrator permissions in Microsoft Azure and Entra ID tenant.

## Step 1/3. Prepare Terraform

> First, set up `azuread` [provider](https://registry.terraform.io/providers/hashicorp/azuread/latest/docs) in your 

Reviewer: ```suggestion
First, set up the `azuread` [provider](https://registry.terraform.io/providers/hashicorp/azuread/latest/docs) in your 
```

working environment from where you want to run the Terraform plan.

> Next, clone the Teleport Entra ID integration example module, which is available in the Teleport Github repository [TODO: update link placeholder](https://github.com/gravitational/teleport/pull/58252).

Reviewer: Making sure the reminder is visible

Author: https://github.com/gravitational/teleport/pull/57835/commits/e43602727ec56e2d395fcbf3a93f30c1afdb1784

> The example provides a configurable Terraform module to setup Entra ID with attributes that are 

Reviewer: ```suggestion
The example provides a configurable Terraform module to set up Entra ID with attributes that are 
```
"setup" is the noun while "set up" is the verb phrase ("set" is the verb).

required for the Teleport Entra ID integration.  

## Step 2/3. Configure and apply

The example Terraform module provides Entra ID configuration option based on the authentication
method you wish to configure for the Teleport Entra ID integration. 

> Setting up Teleport as an ODIC IdP 

Reviewer: Should this be "OIDC"?

> Setting up Teleport as an ODIC IdP 

Reviewer 2: Nit: The early line break looks weird, as if this is an incomplete sentence.

option is the only supported method if you are using Teleport Cloud. If you have a self-hosted 
Teleport Cluster, you must ensure that the Teleport Proxy Service is publicly accessible.

> Alternatively, using system credentials setup is best suited if you have a self-hosted Teleport cluster 

Reviewer: Have we defined system credentials setup before this point?

or when the Teleport Proxy Service is not publicly accessible.  

> At minimum, the Terraform example expects the following input variables.

Reviewer: ```suggestion
At minimum, the Terraform example expects the following input variables:
```
Since we're introducing a list

- `app_name`: Name of the enterprise application that will be created in Entra ID.

> - `proxy_service_address`: Teleport Proxy Service host. E.g., `example.teleport.sh`.

Reviewer: ```suggestion
- `proxy_service_address`: Teleport Proxy Service host, e.g., `example.teleport.sh`.
```

- `certificate_expiry_date`: SAML assertion signing certificate expiry date. 
   E.g., `2026-05-01T01:02:03Z`. Note that an expired certificate will prevent users from 
   signing in to Teleport. If it expires, you will need to update this value in Terraform, 
   as well as in the SAML-based Teleport Auth Connector created for the Entra ID. 

<Tabs>
<TabItem default label="Teleport as an OIDC provider" name="entra-oidc">

Create `tfvars` with your inputs.

```bash
cat > variables.auto.tfvars << EOF
app_name                = "<Var name="Enterprise app name"/>"
proxy_service_address   = "<Var name="Teleport Proxy host"/>"
certificate_expiry_date = "<Var name="SAML certificate expiry"/>"
EOF
```

Applying this plan will perform the following actions:
- Create an enterprise application.
- Set up SAML SSO for Teleport.
- Grant Microsoft Graph API permissions necessary for the Teleport Entra ID integration.
- Setup Teleport as an OIDC-based federated credential provider. 

After you configure `tfvars`, apply the plan.
```bash
$ terraform plan 

$ terraform apply 
```

</TabItem>

> <TabItem label="System Credential" name="entra-system-cred">

Reviewer 2: ```suggestion
<TabItem label="System credential" name="entra-system-cred">
```

You need additional input variables to set up managed identity.
- `use_system_credentials`: Configures Terraform module to plan for resources required 
   for the system credential setup. Value must be `true`.
- `graph_permission_ids`: Permission IDs of the Microsoft Graph API permissions required for the integration:
  - `Application.ReadWrite.OwnedBy`
  - `Group.Read.All`
  - `User.Read.All`
- `managed_id`:  Principal ID of the managed identity which is assigned to the Teleport Auth Service.

The simplest way to retrieve the permission IDs is to run the PowerShell script in your Azure Cloud Shell.

> ```powershell

Reviewer 2: Nit: I'd hide the script under collapsible "details" section.

use_system_credentials   = true
graph_permission_ids    = [<Var name="Permission IDs"/>]
managed_id              = "<Var name="Principal ID of the managed identity"/>"
EOF
```

Applying this plan will perform the following actions:
- Create an enterprise application.
- Set up SAML SSO for Teleport.
- Grant Microsoft Graph API permissions to the managed identity. Permissions 
  are based on the permission IDs which we configured above.

After you configure `tfvars`, apply the plan.
```bash
$ terraform plan 

$ terraform apply 
```

</TabItem>
</Tabs>

Entra ID is now configured with the required attributes needed for the Teleport Entra ID integration.

## Step 3/3. Install the Entra ID plugin

To complete the integration setup, install the Entra ID plugin in Teleport.
The plugin can be installed with the `tctl plugins install entraid` command.

<Tabs>
<TabItem default label="Teleport as an OIDC provider" name="entra-oidc">
```code
$ tctl plugins install entraid \
    --name entra-id-default \
    --auth-connector-name entra-id \
    --default-owner=<Var name="Access List Owner"/> \
    --no-access-graph \
    --manual-setup
```
</TabItem>

> <TabItem label="System Credential" name="entra-system-cred">

Reviewer 2: ```suggestion
<TabItem label="System credential" name="entra-system-cred">
```

    --manual-setup
```
</TabItem>
</Tabs>

- `--name` flag specifies the resource name of the Entra ID plugin.
- `--auth-connector-name` flag specifies the name of the auth connector this integration will create.
- `--default-owner` flag specifies default owners for the Access Lists that will be created
in Teleport based on the groups imported from the Entra ID.
- `--no-access-graph` flag specifies Identity Security integration is to be skipped.
- `--use-system-credentials` flag specifies to use system credential configured using Managed Identity. Use this flag only 
if you with to set up system credential based authentication.
- `--manual-setup` flag specifies a manual Entra ID configuration is selected by the user.

`tctl` will then prompt for an Entra ID tenant ID and an application ID of the enterprise 
application created by using Terraform in step 2.

After you enter these values, the installation command will create a SAML Auth Connector in Teleport, along with the 
plugin service that imports users and groups from Entra ID to Teleport.

## Next steps

- Take a deeper look into setting up [Entra ID auth connector](../../zero-trust-access/sso/azuread.mdx). 

> - Learn more about [Access List](../access-lists/access-lists.mdx) management.

Reviewer 2: I would replace this one with the Entra access control guide you just wrote.

---
title:  Configure Entra ID integration using Terraform
description: Describes how to set up Teleport Entra ID integration using Terraform
sidebar_position: 3
labels:
 - how-to
 - identity-governance
 - terraform
---

> This guide shows Entra ID integration configuration using Terraform and `tctl`.

Reviewer: I would provide some context for when a user would want to follow this guide. Is this an all-Terraform alternative to the guided flow?

  display_name = "example"

  # These resources will be updated later and should not 
  # be overridden.
  lifecycle {
    ignore_changes = [
      identifier_uris,
      required_resource_access,
    ]
  }

  feature_tags {
    enterprise            = true
    custom_single_sign_on = true
  }
}
```

## Step 2/4. Configure SAML SSO

Configure Entra ID enterprise application as an SAML SSO provider for Teleport.
```hcl
locals {
  # Enter your Teleport cluster address. 
  # E.g. format of the expected value: https://example.teleport.sh/v1/webapi/saml/acs/entra-id
  teleport_saml_entity_id = "https://<Var name="example.teleport.sh"/>/v1/webapi/saml/acs/entra-id"
}

resource "azuread_service_principal" "example_sp" {
  client_id = azuread_application.example.client_id

  preferred_single_sign_on_mode = "saml"

  login_url = local.teleport_saml_entity_id

  feature_tags {
    enterprise            = true
    custom_single_sign_on = true
  }
}
```

Configure a certificate to be used for SAML assertion signing.
```hcl
resource "azuread_service_principal_token_signing_certificate" "example_saml_cert" {
  service_principal_id = azuread_service_principal.example_sp.id

>   display_name         = "CN=teleport-tf-example-sso"

Reviewer 2: What is the `teleport-tf-example-sso` here? Is this something users would need to override?

>   end_date             = "2026-05-01T01:02:03Z"

Author: I've updated the terraform example and this guide to mention risk of expired certificate. Though I could not follow how we can automatically tie and update this in the auth connector.

  # Object ID of the MS graph API application. 
  resource_object_id  = data.azuread_service_principal.graph_api.object_id
}

resource "azuread_app_role_assignment" "example_permission_group" {
  # ID of the API permission retrieved using PowerShell. 
  app_role_id         = "<Var name="group-uuid"/>"
  # Object ID of the system assigned Managed identity
  principal_object_id = azurerm_virtual_machine.my_vm.identity[0].principal_id
  # Object ID of the MS graph API application. 
  resource_object_id  = data.azuread_service_principal.graph_api.object_id
}

resource "azuread_app_role_assignment" "example_permission_user" {
  # ID of the API permission retrieved using PowerShell. 
  app_role_id         = "<Var name="user-uuid"/>"
  # Object ID of the system assigned Managed identity
  principal_object_id = azurerm_virtual_machine.my_vm.identity[0].principal_id
  # Object ID of the MS graph API application. 
  resource_object_id  = data.azuread_service_principal.graph_api.object_id
}
```

</TabItem>
</Tabs>

Entra ID is now configured with the required attributes needed for Teleport Entra ID integration.

## Step 4/4. Install the Entra ID plugin

> Integration can be installed with the `tctl plugins install entraid` command.

Reviewer 2: ```suggestion
Teleport side of the integration can be installed with the `tctl plugins install entraid` command.
```
